### PR TITLE
feat: add ec2_sd_configs in ScrapeConfig

### LINKED ISSUE #5818: Support EC2 service discovery with ScrapeConfig CRD
**What happened?**
We are deploying the additionalscrape config using the CRD and ec2sd and kubernetes-sd config is not working.

**Did you expect to see something different?**
Additionalscrape config should be compatible for all the SD configs.

**How to reproduce it (as minimally and precisely as possible)**:
Create a scrapeconfig yaml with ec2 and kubernetes sd config 
**Environment**

* Prometheus Operator version:* 
0.67.1
    `Insert image tag or Git SHA here`
    
    

* Kubernetes version information:
1.24.0
    `kubectl version`
    

* Kubernetes cluster kind:

    insert how you created your cluster: kops, bootkube, etc.

* Manifests:

```
insert manifests relevant to the issue
```

* Prometheus Operator Logs:

```
insert Prometheus Operator logs relevant to the issue here
```

**Anything else we need to know?**:

### PR DESCRIPTION
**Pull request by slashpai**
Fixes #5818

Not all fields are implemented, current change will support monitoring using API access and secret key

## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

Add EC2 Service Discovery support in ScrapeConfig

This is how I tested the change:

- Created an ec2 instance, added security group to open port 9100, installed node-exporter and run it
- Attached the policy `AmazonEC2ReadOnlyAccess` to IAM credentials ([reference](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-Agent-PrometheusEC2.html#CloudWatch-Agent-PrometheusEC2-configure))
- Created k8s secret with api access key and secret key
- Created ScrapeConfig with EC2SDConfig
  ```yaml
  apiVersion: monitoring.coreos.com/v1alpha1
  kind: ScrapeConfig
  metadata:
    name: scrape-config-ec2
    namespace: default
    labels:
      app.kubernetes.io/name: scrape-config-ec2
  spec:
    ec2SDConfigs:
      - region: us-east-1
        accessKey:
          key: accessKey
          name: aws-api-access
        secretKey:
          key: secretKey
          name: aws-api-access
        port: 9100
    relabelings:
    - sourceLabels: [__meta_ec2_public_ip]
      replacement: ${1}:9100
      targetLabel: __address__
   ```

  **Relabelling was required because only public ip will be accessible to scrape target metrics**

Result:

ec_target

I added unit tests, I am exploring how to do e2e test using [localstack](https://docs.localstack.cloud/user-guide/aws/elastic-compute-cloud/) but that will take some time to explore in free time

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
Add EC2 Service Discovery support in ScrapeConfig
```

### CODE REVIEW

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 1]
```diff
@@ -332,3 +335,29 @@ type DNSSDConfig struct {
...
 }
+
+// EC2SDConfig allow retrieving scrape targets from AWS EC2 instances.
+// The private IP address is used by default, but may be changed to the public IP address with relabeling.
+// The IAM credentials used must have the ec2:DescribeInstances permission to discover scrape targets
+// See https://prometheus.io/docs/prometheus/latest/configuration/configuration/#ec2_sd_config
+// +k8s:openapi-gen=true
+type EC2SDConfig struct {
+	// The AWS region. If blank, the region from the instance metadata is used
+	// +optional
```
Comment (slashpai) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:346]: Since we are not implementing other fields like Profile, I am wondering if we should make this required till we implement those

Comment (ArthurSens) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:346]: I'm not again since changing required fields to optional is not a breaking change :)

File: pkg/prometheus/resource_selector.go [hunk 2]
```diff
@@ -720,6 +720,11 @@ func (rs *ResourceSelector) SelectScrapeConfigs(ctx context.Context, listFn List
 			continue
 		}
 
+		if err = rs.validateEC2DConfigs(ctx, sc); err != nil {
```
Comment (ArthurSens) [inline pkg/prometheus/resource_selector.go:723]:
```suggestion
		if err = rs.validateEC2SDConfigs(ctx, sc); err != nil {
```

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 3]
```diff
@@ -332,3 +335,29 @@ type DNSSDConfig struct {
...
+	Region *string `json:"region"`
+	// AccessKey is the AWS API key. If not specified, the environment variable
+	// `AWS_ACCESS_KEY_ID` is used.
```
[the remark this answers is not in the document]
Comment (slashpai) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:349]: correct, updated :)

File: pkg/prometheus/promcfg.go [hunk 4]
```diff
@@ -2575,6 +2576,65 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
+				},
+			}
+
+			if config.AccessKey != nil && config.SecretKey != nil {
+				value, err := store.GetKey(ctx, sc.GetNamespace(), monitoringv1.SecretOrConfigMap{
+					Secret: config.AccessKey,
+				})
+
+				if err != nil {
+					return cfg, errors.Wrapf(err, "failed to get %s access key %s", config.AccessKey.Name, jobName)
```
[the remark this answers is not in the document]
Comment (simonpasquier) [reply pkg/prometheus/promcfg.go:2596]: It's already tracked in https://github.com/prometheus-operator/prometheus-operator/issues/4682
I try to use `fmt.Errorf` for new code but yes we should completely remove `github.com/pkg/error`.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go (hunk 1 above)
Comment (slashpai) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:346]: I am keeping it as optional itself. I think I would need to make access key also required in that case. 
Open for suggestions how we want to do it

File: pkg/prometheus/promcfg.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (slashpai) [reply pkg/prometheus/promcfg.go]: Updated code to use fmt.Errorf for current change

**Comment by slashpai** [comment]
Please have another look on this :)

File: pkg/prometheus/promcfg.go [hunk 5]
```diff
@@ -2575,6 +2576,65 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
 	}
 
+	// EC2SDConfig
+	if len(sc.Spec.EC2SDConfigs) > 0 {
+		configs := make([][]yaml.MapItem, len(sc.Spec.EC2SDConfigs))
+		for i, config := range sc.Spec.EC2SDConfigs {
+			configs[i] = []yaml.MapItem{
+				{
+					Key:   "region",
+					Value: config.Region,
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2586]: I think that we need to check for nil.

File: pkg/prometheus/promcfg_test.go [hunk 6]
```diff
@@ -8245,6 +8245,88 @@ func TestScrapeConfigSpecConfigWithConsulSD(t *testing.T) {
...
+				"secretKey": []byte("secret-key"),
+			},
+		},
+	)
+	for _, tc := range []struct {
+		name   string
+		scSpec monitoringv1alpha1.ScrapeConfigSpec
+		golden string
+	}{
+		{
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg_test.go:8266]: Can you add a test with an empty config + another one with an invalid config?

File: pkg/prometheus/promcfg.go (hunk 5 above)
Comment (slashpai) [reply pkg/prometheus/promcfg.go:2586]: per prometheus region can be empty but in that case other fields are expected to get instance metadata. So I had this thought https://github.com/prometheus-operator/prometheus-operator/pull/5902#discussion_r1320312099 whether to make this required field till we implement/plan to support other methods other than with api access keys

**Comment by slashpai** [comment]
addressed comments

**Review by ArthurSens — APPROVED** [review]
